// @uiw/react-heat-map/lib/LabelsWeek.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.textStyle = exports.LabelsWeek = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _react = _interopRequireWildcard(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
var textStyle = exports.textStyle = {
  textAnchor: 'middle',
  fontSize: 'inherit',
  fill: 'currentColor'
};
var LabelsWeek = exports.LabelsWeek = function LabelsWeek(_ref) {
  var _ref$weekLabels = _ref.weekLabels,
    weekLabels = _ref$weekLabels === void 0 ? [] : _ref$weekLabels,
    _ref$rectSize = _ref.rectSize,
    rectSize = _ref$rectSize === void 0 ? 0 : _ref$rectSize,
    _ref$topPad = _ref.topPad,
    topPad = _ref$topPad === void 0 ? 0 : _ref$topPad,
    _ref$space = _ref.space,
    space = _ref$space === void 0 ? 0 : _ref$space;
  return (0, _react.useMemo)(function () {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_react.Fragment, {
      children: (0, _toConsumableArray2["default"])(Array(7)).map(function (_, idx) {
        if (weekLabels && weekLabels[idx]) {
          return /*#__PURE__*/(0, _jsxRuntime.jsx)("text", {
            className: "w-heatmap-week",
            x: 15,
            y: topPad,
            dy: (idx + 1) * (rectSize + space) - 5,
            style: textStyle,
            children: weekLabels[idx]
          }, idx);
        }
        return null;
      })
    });
  }, [rectSize, space, topPad, weekLabels]);
};
